Service::Y2R: Support for grayscale decoding of specific formats

yuriks opened the pull request:
Implements unrotated planar YUV 4:2:0 -> RGB24 conversions in Y2R.
Currently only the Y (luma) channel is used, so the results don't
contain color. This will be added in a later PR at some point.

This is enough to get all currently know Moflex videos to decode. (Some
don't display on-screen due to seemingly unrelated reasons.)

Thanks to @archshift for doing the initial implementation which I
cleaned up and then fixed the 8x8 block mode.

Subv commented:
I think you should reset those variables to some default value when the module is initialized

yuriks commented:
As far as I've seen, all games set all the registers before a conversion, but fair point.

yuriks commented:
I moved all the conversion parameters into a struct so that it can be easily memsetted. This also gives me an excuse to move all params used in the conversion to local variables, which should help compile analysis and help generate faster code. While at it I also changed all variables to `size_t` which also helps generate better code in x64 (due to `int` being different from the native register size).

Subv commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -2,84 +2,307 @@
 // Licensed under GPLv2 or any later version
```
I'd say it's fair to change the copyright year to 2015 now

bunnei replied:
+1

archshift commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -2,84 +2,307 @@
+    ASSERT(conversion_params.output_format == OutputFormat::OUTPUT_RGB_24);
+    ASSERT(conversion_params.rotation == Rotation::ROTATION_NONE);
+
+    const int bpp = 3;
```
Add a TODO here

yuriks commented:
Updated.

archshift commented:
LGTM

Subv commented:
LGTM

bunnei commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -1,85 +1,308 @@
 
 namespace Y2R_U {
 
+enum InputFormat {
```
Can you change all of these to `enum class ClassName : u32 {`, then get rid of the duplicate prefixes (e.g. `INPUT_`, `OUTPUT_`, `ROTATION_`, ...)

bunnei replied:
Also, some documentation would be nice (and/or in the struct below)

yuriks replied:
These enums were taken from the BPB symbol table. Still change them?

bunnei commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -1,85 +1,308 @@
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    conversion_params.output_format = static_cast<OutputFormat>(cmd_buff[1]);
+    LOG_DEBUG(Service, "Y2R OutputFormat: %u", conversion_params.output_format);
```
Can we change this `LOG_DEBUG` (and the one's below to be consistent) with how we log service routines? E.g. replace the `Y2R` (we know what service this called from..) with `called`.

archshift replied:
I also just noticed this, but `Service` should instead be some specific service log format, like `Service_Y2R`

bunnei commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -1,85 +1,308 @@
+    conversion_params.srcY_transfer_unit = cmd_buff[3];
+    conversion_params.srcY_stride = cmd_buff[4];
+    u32 src_process_handle = cmd_buff[6];
+    LOG_DEBUG(Service, "called! image_size: 0x%08X, transfer_unit: %hu, transfer_stride: %hu, "
```
`called!`? Ditto format with above^

bunnei commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -1,85 +1,308 @@
+    {
+        const size_t input_lines = conversion_params.input_lines;
+        const size_t input_line_width = conversion_params.input_line_width;
+        const size_t srcY_stride = conversion_params.srcY_stride;
```
`srcY_stride` ?

bunnei commented on src/core/hle/service/y2r_u.cpp:
```diff
@@ -1,85 +1,308 @@
+    VAddr srcY_address;
+    u32 srcY_image_size;
+    u16 srcY_transfer_unit;
+    u16 srcY_stride;
```
`srcY` "is not the preferred nomenclature, dude"

yuriks replied:
"Y" here stands for the luma channel in YUV, which is always written like that. I'll keep it and add comments.

bunnei commented:
Just come stylistic comments, please address & squash

yuriks commented:
Updated

bunnei commented:
LGTM
